prototype_source/nestedtensor.py

```diff
@@ -368,124 +264,100 @@ def zipf_sentence_lengths(alpha: float, batch_size: int) -> np.ndarray:
+print("nested and padded calculations differ by", (jagged_to_padded(output_nested, 0.0) - output_padded).abs().max().item())
+print("nested tensor multi-head attention takes", time_nested, "seconds")
+print("padded tensor multi-head attention takes", time_padded, "seconds")
+
+# warm up compile first...
+compiled_mha = torch.compile(mha)
+compiled_mha(query, key, value)
+# ...now benchmark
+compiled_output_nested, compiled_time_nested = benchmark(
+    compiled_mha, query, key, value)
+
+# warm up compile first...
+compiled_mha(padded_query, padded_key, padded_value)
+# ...now benchmark
+compiled_output_padded, compiled_time_padded = benchmark(
+    compiled_mha, padded_query, padded_key, padded_value)
+
+# padding-specific step: remove output projection bias from padded entries for fair comparison
+for i, entry_length in enumerate(sentence_lengths):
+    compiled_output_padded[i, entry_length:] = 0.0
+
+print("=== with torch.compile ===")
+print("nested and padded calculations differ by", (jagged_to_padded(compiled_output_nested, 0.0) - compiled_output_padded).abs().max().item())
+print("nested tensor multi-head attention takes", compiled_time_nested, "seconds")
+print("padded tensor multi-head attention takes", compiled_time_padded, "seconds")
 
 ######################################################################
-# If we set need_weights to False this will enable the fast path in the library.
-# Under the hood this will call _scaled_dot_product_attention. If your tensors
-# are on CUDA, than a fused, efficient attention kernel will be used. For
-# more detailed performance characteristics look at the benchmark in
-# pytorch/benchmarks/transformer/sdp.py
-
-with torch.inference_mode():
-    t0 = timeit.default_timer()
-    out_lib, out_lib_weights = mha_lib(query, query, query, need_weights=False)
-
-    t1 = timeit.default_timer()
-    padded_out = mha_padded(
-        padded_query, padded_query, padded_query, nheads,
-        attn_mask_q, attn_mask_q,
-        W_q, W_k, W_v, W_out,
-        b_q=b_q, b_k=b_k, b_v=b_v, b_out=b_out,
-        dropout_p=dropout_p)
-    t2 = timeit.default_timer()
-
-nested_time = t1 - t0
-padded_time = t2 - t1
-print("Nested and padded calculations differ by", (torch.nested.to_padded_tensor(out_lib, 0.0) - padded_out).abs().max().item())
-print("Nested library multi-head attention takes", nested_time, "seconds")
-print("Padded tensor multi-head attention takes", padded_time, "seconds")
-print(f"Nested Speedup: {padded_time / nested_time:.3f}")
+# Note that without torch.compile, the overhead of the python subclass nested tensor
+# can make it slower than the equivalent computation on padded tensors. However, once
+# torch.compile is enabled, operating on nested tensors gives a multiple x speedup.
+# Avoiding wasted computation on padding becomes only more valuable as the percentage
+# of padding in the batch increases.
+print(f"Nested speedup: {compiled_time_padded / compiled_time_nested:.3f}")
```

Comment:
Need to add a Conclusion here:
```suggestion
print(f"Nested speedup: {compiled_time_padded / compiled_time_nested:.3f}")

Conclusion
-------------

In this tutorial, we have learned how to perform basic operations with nested tensors. 
```

Need to add a lit of resources that the user can read:

For more information about nested tensors, see:

* Link1
* Link2 
* Link3